ssl/ssl_cert_comp.c

```diff
@@ -0,0 +1,464 @@
+{
+    OSSL_COMP_CERT *ret = NULL;
+
+    if (!ossl_comp_has_alg(alg)
+            || data == NULL
+            || (ret = OPENSSL_zalloc(sizeof(*ret))) == NULL
+            || (ret->lock = CRYPTO_THREAD_lock_new()) == NULL)
+        goto err;
+
+    ret->references = 1;
+    ret->data = data;
+    ret->len = len;
+    ret->orig_len = orig_len;
+    ret->alg = alg;
+    return ret;
+ err:
+    ERR_raise(ERR_LIB_SSL, ERR_R_MALLOC_FAILURE);
+    OPENSSL_free(data);
+    OPENSSL_free(ret);
+    return NULL;
+}
+
+__owur static OSSL_COMP_CERT *OSSL_COMP_CERT_compressed_data(unsigned char *data, size_t len, size_t orig_len, int alg)
+{
+    return OSSL_COMP_CERT_new(OPENSSL_memdup(data, len), len, orig_len, alg);
+}
+
+__owur static OSSL_COMP_CERT *OSSL_COMP_CERT_uncompressed_data(unsigned char *data, size_t len, int alg)
+{
+    OSSL_COMP_CERT *ret = NULL;
+    size_t max_length;
+    int comp_length;
+    COMP_METHOD *method;
+    unsigned char *comp_data = NULL;
+    COMP_CTX *comp_ctx = NULL;
+
+    switch (alg) {
+    case TLSEXT_comp_cert_brotli:
+        method = COMP_brotli_oneshot();
+        break;
+    case TLSEXT_comp_cert_zlib:
+        method = COMP_zlib();
+        break;
+    case TLSEXT_comp_cert_zstd:
+        method = COMP_zstd_oneshot();
+        break;
+    default:
+        goto err;
+    }
+
+    if ((max_length = ossl_calculate_comp_expansion(alg, len)) == 0
+          || method == NULL
+          || (comp_ctx = COMP_CTX_new(method)) == NULL
+          || (comp_data = OPENSSL_zalloc(max_length)) == NULL)
+        goto err;
+
+    comp_length = COMP_compress_block(comp_ctx, comp_data, max_length, data, len);
+    if (comp_length <= 0)
+        goto err;
+
+    ret = OSSL_COMP_CERT_new(comp_data, comp_length, len, alg);
+    comp_data = NULL;
+
+ err:
+    OPENSSL_free(comp_data);
+    COMP_CTX_free(comp_ctx);
+    return ret;
+}
+
+void OSSL_COMP_CERT_free(OSSL_COMP_CERT *cc)
+{
+    int i;
+
+    if (cc == NULL)
+        return;
+
+    CRYPTO_DOWN_REF(&cc->references, &i, cc->lock);
+    REF_PRINT_COUNT("OSSL_COMP_CERT", cc);
+    if (i > 0)
+        return;
+    REF_ASSERT_ISNT(i < 0);
+
+    OPENSSL_free(cc->data);
+    CRYPTO_THREAD_lock_free(cc->lock);
+    OPENSSL_free(cc);
+}
+int OSSL_COMP_CERT_up_ref(OSSL_COMP_CERT *cc)
+{
+    int i;
+
+    if (CRYPTO_UP_REF(&cc->references, &i, cc->lock) <= 0)
+        return 0;
+
+    REF_PRINT_COUNT("OSSL_COMP_CERT", cc);
+    REF_ASSERT_ISNT(i < 2);
+    return ((i > 1) ? 1 : 0);
+}
+
+static int ssl_set_cert_comp_pref(int *prefs, int *algs, size_t len)
+{
+    int j = 0;
+    int i;
+    int found = 0;
+    int already_set[TLSEXT_comp_cert_limit];
+    int tmp_prefs[TLSEXT_comp_cert_limit];
+    int num_algs;
+
+    /* clear all algorithms */
+    if (len == 0 || algs == NULL) {
+        memset(prefs, 0, sizeof(tmp_prefs));
+        return 1;
+    }
+
+    num_algs = len / sizeof(*algs);
```

Comment:
Hmm, why is the length specified in bytes? The documentation is also unclear:
```
The B<algs> argument is an array
of algorithms, and B<length> is size of the B<algs> array.
```

I would prefer it if the API were to be changed to specify the number of array elements.

Reply:
That's how it was originally, but it was changed suggested by @mattcaswell. I can add an "in bytes" to this to make it clear.

Reply:
I think what I actually asked for here was for this to be a `size_t`. I would also prefer the size to be the number of elements rather the number of bytes.

Reply:
`size_t` implies bytes, IMHO... but I really don't care *that* much. I will change to be the number of elements in the documentation and the code.

Reply:
> size_t implies bytes, IMHO

Not to me. It implies a size. E.g. see the `nmemb` param to calloc:

````
void *calloc(size_t nmemb, size_t size);
````

Reply:
That's fair. I've put it to be the number of elements.